Adapted flask restx for newer dependencies and python 3 10

Andrei2D opened the pull request.

sinescu-andreea committed:
Add description and enum values for body parameter

sinescu-andreea committed:
Send kwargs to be able to send add_specs parameter to disable or not the swagger documentation

sinescu-andreea committed:
Mark the required parameters from body

Commit:
Added in utils the function removed from wergzeug

Commit:
Updated:
- requirements versions
- python-compatible versions
- package version

Commit:
Updated request parser for json and get_json

Commit:
Adapted tests that break due to dependencies update

Andrei2D commented on flask_restx/reqparse.py:
```diff
@@ -138,22 +138,24 @@ def source(self, request):
+
+        values = MultiDict()
+        for l in _args_location:
+            if l == 'json' and hasattr(request, 'get_json'):
```
Normally, `request.json` is attempted by default (property calling method `get_json`) but method's old behavior is now similar to having to pass `silent=True` parameter to the current.

Long story short: newer werkzeug package raises errors when previously it would not.

Andrei2D replied:
Undid these changes and updated tests to include proper "Content-Type" header when using json body.

Andrei2D commented on flask_restx/utils.py:
```diff
@@ -122,3 +137,38 @@ def unpack(response, default_code=HTTPStatus.OK):
         raise ValueError("Too many response values")
+
+
+def parse_rule(rule: str) -> t.Iterator[t.Tuple[t.Optional[str], t.Optional[str], str]]:
```
Function entirely removed from `werkzeug` on newer versions

andreeasinescusst replied:
But doesn't they have another one to do something like this?

Andrei2D replied:
[It does](https://github.com/pallets/werkzeug/blob/2.2.x/src/werkzeug/routing/rules.py#L584) but it seemed more cumbersome than [the initial](https://github.com/pallets/werkzeug/blob/2.0.x/src/werkzeug/routing.py#L215) function.

Gonna be honest: haven't looked into this new module much as I found importing the function faster than understanding the new logic and adapting (with the possibility of it not being equivalent). If needed be, I can do that,

Andrei2D commented on tests/test_logging.py:
```diff
@@ -86,7 +86,9 @@ def get(self):
-        assert len(matching) == 1
+        # NOTE: previously set to 1; 0 is accurate as debug level is lower than
+        # info level, which is set to the root logger
+        assert len(matching) == 0
```
This test is (for some reason) related to `bleach` dependency (as changing it's version, to support higher pyhton version, made this test fail).

vlungusst replied:
did you try printing the output before your changes and after to see what differs? this looks like a domain name server connectivity check

Andrei2D replied:
Updated this test.

The issue was pytest's `LogCaptureFixture` and the fact that when settings the level to a logger trough an instance of that class it would update the log level of that instance's handler (in charge of intercepting messages for assertions). In turn, trying to send a debug log when the level is info would fail.

Andrei2D commented on tests/legacy/test_api_legacy.py:
```diff
@@ -379,7 +379,10 @@ def get(self):
-        assert resp.headers["Location"] == "http://localhost/"
+
+        # FIXME: did not found where base url is filled in location headers
+        # assert resp.headers["Location"] == "http://localhost/"
```
This used not to break for the builds on python 3.5 and 3.6 (with the older requirements).

vlungusst replied:
what is resp.headers["Location"] value now?

Andrei2D replied:
Umm... line 385:
`assert resp.headers["Location"] == "/"`

Andrei2D replied:
Not sure what else to add to this thread.

Commit:
Request parse updates
* undid Argument.source function to its old version
* left only args params as default argument location
* enforced proper header use and arguments declaration in unit tests

Commit:
Updated package version

Commit:
Removed ossaudit pipeline

Commit:
Changed version number to the right increment

Commit:
Updated pipeline to have tests run without checks

Commit:
Updated git versions

Andrei2D commented:
The last pipeline issue is related to the fact that tests on master no longer work due to an import error (which was fixed in this PR).

vlungusst commented on .github/workflows/test.yml:
```diff
@@ -10,45 +10,44 @@ on:
+  #         echo ${RES}
   unit-tests:
-    needs: checks
     runs-on: ubuntu-latest
```
I'd specify a version of ubuntu we know works (like 20.04); it shouldn't influence this particular repo, but we've had issues with 22.04 and mongo as the package maintainer was behind and didn't have packages set up for it

vlungusst commented on examples/zoo_app/requirements.txt:
```diff
@@ -3,12 +3,12 @@ attrs==21.2.0
 Flask==1.1.4
 flask-restx==0.5.1
-itsdangerous==1.1.0
+itsdangerous==2.0.1
```
just learned about this, this package seems awesome :)))

Commit:
Updated ubuntu versions used in pipelines

vlungusst commented on flask_restx/utils.py:
```diff
@@ -122,3 +137,38 @@ def unpack(response, default_code=HTTPStatus.OK):
+    """
+    pos = 0
+    end = len(rule)
+    do_match = _rule_re.match
```
why not define rule_re here (afaics it's only used here), so that the werkzeug method is self contained)?

andreeasinescusst commented on .github/workflows/test.yml:
```diff
@@ -10,45 +10,44 @@ on:
-            exit 1
-          fi
-          echo ${RES}
+  # checks:
```
Why do we commented this?

Andrei2D replied:
It was checking vulnerabilities in the packages that the library is using and was failing as it detected some. It triggered for Flask, setuptools and around 3 more other.

However, there was no alternative version that we could've put. [Flask](https://ossindex.sonatype.org/component/pkg:pypi/flask) for instance seems to have a cookie-related vulnerability (that we're not using anyway) in all its versions. The same story with [setuptools](https://ossindex.sonatype.org/component/pkg:pypi/setuptools).

Therefore, after talking with Vali about it, we agreed to comment it (and keep it commented for later).

andreeasinescusst commented on .github/workflows/release.yml:
```diff
@@ -5,7 +5,7 @@ on:
 jobs:
   build:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
```
Check what failed in checks

Andrei2D replied:
It seemed like it attempted to run the master but the master would install latest werkzeug version and would fail importing the function that was added in this PR into `utils.py`.
![image](https://user-images.githubusercontent.com/32478989/186588107-807f3c81-b151-4bab-847d-aea3c0227364.png)


Not sure what to do about it, However, after this PR, this issue (in particular, should not happen anymore).

Commit:
Fixed log related test and updated parse_rule

The isse with the test was that the fixture pytest had for helping the tests was not catching all messages that logs were allowed to log.

Commit:
Merge branch 'bugfix/help-and-choices-are-ignored-for-body-params' of github.com:andreeasinescusst/flask-restx into task/python-3-10
